pubsub/google/cloud/pubsub_v1/futures.py

```diff
@@ -0,0 +1,176 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import threading
+
+import google.api_core.future
+from google.cloud.pubsub_v1.publisher import exceptions
+
+
+class Future(google.api_core.future.Future):
+    """Encapsulation of the asynchronous execution of an action.
+
+    This object is returned from asychronous Pub/Sub calls, and is the
+    interface to determine the status of those calls.
+
+    This object should not be created directly, but is returned by other
+    methods in this library.
+    """
+    _SENTINEL = object()
+
+    def __init__(self):
+        self._result = self._SENTINEL
+        self._exception = self._SENTINEL
+        self._callbacks = []
+        self._completed = threading.Event()
+
+    def cancel(self):
+        """Actions in Pub/Sub generally may not be canceled.
+
+        This method always returns False.
+        """
+        return False
+
+    def cancelled(self):
+        """Actions in Pub/Sub generally may not be canceled.
+
+        This method always returns False.
+        """
+        return False
+
+    def running(self):
+        """Actions in Pub/Sub generally may not be canceled.
+
+        Returns:
+            bool: ``True`` if this method has not yet completed, or
+                ``False`` if it has completed.
+        """
+        if self.done():
+            return False
+        return True
+
+    def done(self):
+        """Return True if the publish has completed, False otherwise.
+
+        This still returns True in failure cases; checking :meth:`result` or
+        :meth:`exception` is the canonical way to assess success or failure.
+        """
+        return (self._exception is not self._SENTINEL or
+                self._result is not self._SENTINEL)
+
+    def result(self, timeout=None):
+        """Return the message ID, or raise an exception.
+
+        This blocks until the message has successfully been published, and
+        returns the message ID.
+
+        Args:
+            timeout (Union[int, float]): The number of seconds before this call
+                times out and raises TimeoutError.
+
+        Returns:
+            str: The message ID.
+
+        Raises:
+            ~.pubsub_v1.TimeoutError: If the request times out.
+            Exception: For undefined exceptions in the underlying
+                call execution.
+        """
+        # Attempt to get the exception if there is one.
+        # If there is not one, then we know everything worked, and we can
+        # return an appropriate value.
+        err = self.exception(timeout=timeout)
+        if err is None:
+            return self._result
+        raise err
+
+    def exception(self, timeout=None, _wait=1):
```

Comment:
what's the purpose of `_wait` here?

Reply:
Oops, old argument that persisted through a previous refactor (this file is just moved from another location). I will remove it.